src/tree/identity_tree.rs

```diff
@@ -0,0 +1,611 @@
+                Ok(Some(InclusionProof::new(self.tree.root(), proof)))
+            } else {
+                let proof = self.construct_proof_from_root(*leaf_idx, root)?;
+                Ok(Some(InclusionProof::new(root.hash, proof)))
+            }
+        } else {
+            let proof = self.tree.proof(*leaf_idx as usize);
+            Ok(Some(InclusionProof::new(self.tree.root(), proof)))
+        }
+    }
+
+    pub fn construct_proof_from_root(
+        &self,
+        leaf_idx: u32,
+        root: &Root,
+    ) -> eyre::Result<Proof> {
+        let updates = self
+            .tree_updates
+            .get(root)
+            .ok_or_eyre("Could not find root in tree updates")?;
+
+        let mut node_idx = leaf_to_storage_idx(leaf_idx, self.tree.depth());
+
+        let mut proof: Vec<Branch<Hash>> = vec![];
+
+        while node_idx > 0 {
+            let sibling_idx = if node_idx % 2 == 0 {
+                node_idx - 1
+            } else {
+                node_idx + 1
+            };
+
+            let sibling = updates
+                .get(&sibling_idx)
+                .copied()
+                .or_else(|| {
+                    let (depth, offset) =
+                        storage_idx_to_coords(sibling_idx as usize);
+                    Some(self.tree.get_node(depth, offset))
+                })
+                .expect("Could not find node in tree");
+
+            proof.push(if node_idx % 2 == 0 {
+                Branch::Right(sibling)
+            } else {
+                Branch::Left(sibling)
+            });
+
+            node_idx = (node_idx - 1) / 2;
+        }
+
+        Ok(semaphore::merkle_tree::Proof(proof))
+    }
+
+    pub fn insert(&mut self, index: u32, leaf: Hash) -> eyre::Result<()> {
+        // Check if the leaf laready exists
+        if self.leaves.contains_key(&leaf) {
+            return Err(eyre!("Leaf already exists"));
+        }
+
+        self.leaves.insert(leaf, index);
+
+        // We can expect here because the `reallocate` implementation for Vec<H::Hash> as DynamicTreeStorage does not fail
+        self.tree.push(leaf).expect("Failed to insert into tree");
+
+        Ok(())
+    }
+
+    pub fn remove(&mut self, index: usize) {
+        let leaf = self.tree.get_leaf(index);
+        self.leaves.remove(&leaf);
+        self.tree.set_leaf(index, Hash::ZERO);
+    }
+
+    // Appends new leaf updates and newly calculated intermediate nodes to the tree updates
+    pub fn append_updates(&mut self, root: Root, leaf_updates: LeafUpdates) {
+        // Update leaves
+        match leaf_updates {
+            LeafUpdates::Insert(ref updates) => {
+                for (idx, val) in updates.iter() {
+                    self.leaves.insert(*val, *idx);
+                }
+            }
+            LeafUpdates::Delete(ref updates) => {
+                for (_, val) in updates.iter() {
+                    self.leaves.remove(val);
+                }
+            }
+        }
```

Comment:
Not sure why I feel this way, but `ref` feels like a leftover from the days before match ergonomics were a thing.

```suggestion
        match &leaf_updates {
            LeafUpdates::Insert(updates) => {
                for (idx, val) in updates.iter() {
                    self.leaves.insert(*val, *idx);
                }
            }
            LeafUpdates::Delete(updates) => {
                for (_, val) in updates.iter() {
                    self.leaves.remove(val);
                }
            }
        }
```